lib/DxilDia/DxcPixLiveVariables.cpp

```diff
@@ -0,0 +1,268 @@
+  auto *Expression = DbgDeclare->getExpression();
+
+  if (Address == nullptr || Variable == nullptr || Expression == nullptr)
+  {
+    return;
+  }
+
+  auto* AddressAsAlloca = llvm::dyn_cast<llvm::AllocaInst>(Address);
+  if (AddressAsAlloca == nullptr)
+  {
+      return;
+  }
+
+  auto *S = Variable->getScope();
+  if (S == nullptr)
+  {
+    return;
+  }
+
+  std::unique_ptr<FragmentIterator> Iter = FragmentIterator::Create(
+      DbgDeclare,
+      m_pModule->getDataLayout(),
+      AddressAsAlloca,
+      Expression);
+
+  if (!Iter)
+  {
+    // FragmentIterator creation failure, this skip this var.
+    return;
+  }
+
+  const unsigned FragmentSizeInBits = Iter->FragmentSizeInBits();
+
+  unsigned FragmentIndex;
+  while (Iter->Next(&FragmentIndex))
+  {
+    const unsigned FragmentOffsetInBits =
+        Iter->CurrOffsetInBits();
+
+    VariableInfo* VarInfo = AssignValueToOffset(
+        &m_LiveVarsDbgDeclare[S],
+        Variable,
+        Address,
+        FragmentIndex,
+        FragmentOffsetInBits);
+
+    ValidateDbgDeclare(
+        VarInfo,
+        FragmentSizeInBits,
+        FragmentOffsetInBits);
+  }
+}
+
+dxil_debug_info::VariableInfo* dxil_debug_info::LiveVariables::Impl::AssignValueToOffset(
+    VariableInfoMap *VarInfoMap,
+    llvm::DIVariable *Variable,
+    llvm::Value *Address,
+    unsigned FragmentIndex,
+    unsigned FragmentOffsetInBits
+)
+{
+  // FragmentIndex is the index within the alloca'd value
+  // FragmentOffsetInBits is the offset within the HLSL variable
+  // that maps to Address[FragmentIndex]
+  auto it = VarInfoMap->find(Variable);
+  if (it == VarInfoMap->end())
+  {
+    auto InsertIt = VarInfoMap->emplace(
+        Variable,
+        std::make_unique<VariableInfo>(Variable));
+    assert(InsertIt.second);
+    it = InsertIt.first;
+  }
+  auto *VarInfo = it->second.get();
+  auto &FragmentLocation =
+      VarInfo->m_ValueLocationMap[FragmentOffsetInBits];
+  FragmentLocation.m_V = Address;
+  FragmentLocation.m_Offset = FragmentIndex;
+
+  return VarInfo;
+}
+
+dxil_debug_info::LiveVariables::LiveVariables() = default;
+
+dxil_debug_info::LiveVariables::~LiveVariables() = default;
+
+HRESULT dxil_debug_info::LiveVariables::Init(DxcPixDxilDebugInfo *pDxilDebugInfo) {
+  Clear();
+  m_pImpl->Init(pDxilDebugInfo->GetMallocNoRef(), pDxilDebugInfo, pDxilDebugInfo->GetModuleRef());
+  return S_OK;
+}
+
+void dxil_debug_info::LiveVariables::Clear() {
+  m_pImpl.reset(new dxil_debug_info::LiveVariables::Impl());
+}
+
+HRESULT dxil_debug_info::LiveVariables::GetLiveVariablesAtInstruction(
+  llvm::Instruction *IP,
+  IDxcPixDxilLiveVariables **ppResult) const {
+  DXASSERT(IP != nullptr, "else IP should not be nullptr");
+  DXASSERT(ppResult != nullptr, "else Result should not be nullptr");
+
+  std::vector<const VariableInfo *> LiveVars;
+  std::set<std::string> LiveVarsName;
+
+  const llvm::DebugLoc &DL = IP->getDebugLoc();
+
+  if (!DL)
+  {
+    return S_FALSE;
```

Comment:
So this is an evil S_FALSE. If the caller used if(SUCCEEDED on this method, they'd crash trying to use *ppResult.